rust/wasmtime_bindings/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+        fn pop_val(self: &mut ValVec) -> Result<Box<Val>>;
+        fn get_val_vec() -> Result<Box<ValVec>>;
+
+        // type and methods for returning and executing a WebAssembly function asynchronously
+        type Fut<'a>;
+        unsafe fn resume<'a>(self: &mut Fut<'a>) -> Result<bool>;
+        unsafe fn get_func_future<'a>(
+            store: &'a mut Store,
+            func: &'a Func,
+            args: &'a ValVec,
+            rets: &'a mut ValVec,
+        ) -> Result<Box<Fut<'a>>>;
+    }
+}
+
+pub struct Instance {
+    wasmtime_instance: wasmtime::Instance,
+}
+
+fn create_instance(engine: &Engine, module: &Module, store: &mut Store) -> Result<Box<Instance>> {
+    let mut linker = wasmtime::Linker::new(&engine.wasmtime_engine);
+    if let Err(e) = wasmtime_wasi::tokio::add_to_linker(&mut linker, |s| s) {
+        return Err(anyhow!("Failed to add wasi to linker: {:?}", e));
+    }
+    let mut inst_fut =
+        Box::pin(linker.instantiate_async(&mut store.wasmtime_store, &module.wasmtime_module));
+    let waker = futures::task::noop_waker_ref();
+    let mut ctx = Context::from_waker(&waker);
+
+    loop {
+        match inst_fut.as_mut().poll(&mut ctx) {
+            Poll::<Result<wasmtime::Instance>>::Pending => {}
+            Poll::<Result<wasmtime::Instance>>::Ready(Ok(inst)) => {
+                return Ok(Box::new(Instance {
+                    wasmtime_instance: inst,
+                }));
+            }
+            Poll::<Result<wasmtime::Instance>>::Ready(Err(e)) => {
+                return Err(anyhow!("Failed to instantiate module: {:?}", e));
+            }
+        }
+    }
+}
+
+pub struct Module {
+    wasmtime_module: wasmtime::Module,
+}
+
+fn create_module(engine: &mut Engine, script: &str) -> Result<Box<Module>> {
+    match wasmtime::Module::new(&engine.wasmtime_engine, script) {
+        Ok(module) => Ok(Box::new(Module {
+            wasmtime_module: module,
+        })),
+        Err(e) => Err(anyhow!("Compilation failed: {}", e)),
+    }
+}
+
+pub struct Store {
+    wasmtime_store: wasmtime::Store<wasmtime_wasi::WasiCtx>,
+}
+
+fn create_store(engine: &mut Engine) -> Result<Box<Store>> {
+    let wasi = wasmtime_wasi::tokio::WasiCtxBuilder::new().build();
+    let mut store = wasmtime::Store::new(&mut engine.wasmtime_engine, wasi);
+    // TODO: make this configurable
+    store.out_of_fuel_async_yield(10000, 4096);
+    Ok(Box::new(Store {
+        wasmtime_store: store,
+    }))
+}
+
+pub struct Memory {
+    wasmtime_memory: wasmtime::Memory,
+}
+
+impl Memory {
+    fn data(self: &Memory, store: &Store) -> *mut u8 {
+        self.wasmtime_memory.data_ptr(&store.wasmtime_store)
+    }
+    fn size(self: &Memory, store: &Store) -> u64 {
+        self.wasmtime_memory.size(&store.wasmtime_store)
+    }
+    fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64> {
+        self.wasmtime_memory.grow(&mut store.wasmtime_store, delta)
```

Comment:
What does wasmtime use to allocate memory?

Reply:
(Need to understand if it will be accounted for in seastar memory management)

Reply:
I skimmed through wasmtime code and it indeed mmaps a lot. Some of it is actual file-backed mapping, but often it's used as glorified malloc with MAP_ANONYMOUS. 

We could technically add an opt-in Seastar feature which makes extrernal mmap-based libraries more Seastar friendly:
1. Override `mmap` to fall back to original behavior, except when `MAP_ANONYMOUS` is set, in which case we just call aligned_alloc with page size boundary, which effectively gives us pages. Then, we can apply requested protection with a separate `mprotect` call.
2. `munmap` gets a little complicated, because I don't know any reliable way of checking if a page is file-backed or anonymous, except for examining `/proc/self/maps` output. But if we assume that the number of anonymous mappings is relatively small, perhaps adding a hashmap overhead is acceptable if it means that we regain the ability to allocate everything with a Seastar allocated. Then, we could just track our anonymous pages ourselves and override `munmap` behavior to a regular `free` if the page is tracked as "ours".

@avikivity Does it sound too obscure to you, or is it feasible? The main question is how many more mmap-ed libraries we're likely to integrate.

Reply:
Won't file-backed `mmap()` cause blocking memory accesses and stalls?

Reply:
Not just stalls, also OOM since we already allocated all memory for our own use.

Reply:
Overriding ::mmap is likely to end up with a lot of tears, it's too fundamental and too not designed to be overridden.

Can we patch wasmtime (upstream) not to use mmap? It seems like premature optimization. The system memory allocator will call mmap when the conditionals call for it.

Perhaps the issue is the need to make the memory executable later with mprotect() (typical jit operation is to first mmap with PROT_WRITE, then mprotect to PROT_EXEC.

If so, perhaps we can pass callbacks to wasmtime for mmap/mprotect/munmap so we can intercept them. The default callbacks will just call the system mmap/mprotect/munmap.

Reply:
Completely removing `mmap` from wasmtime seems difficult, but maybe it is not essential after all. 
I inspected the wasmtime code and I found that the `mmap`s are used in:
- the default `Linear Memory` allocator - which we have replaced
- the `Pooling Memory Allocator` (where we preallocate all the memory for all the instances that we can later use) - which we do not use
- in the `copy-on-write` implementation of the initialization of `Linear Memories` - which is only used in the before-mentioned `Pooling Memory Allocator`
- for allocation of the stacks for asynchronous calls - this could probably be replaced with `malloc`
- for allocation of the alternate signal stack - this could also probably be replaced with `malloc`
- in compilation of a module from a file - which we currently do not use
- for allocation of for the jit code to be later `mprotect`ed with `PROT_EXEC` - the `mmap`s are all with `MAP_ANONYMOUS` but I'm not sure we can help it (according to the `mprotect` manual `mprotect()` is unspecified if it is applied to a of memory that was not obtained via `mmap()`)

Reply:
> for allocation of for the jit code to be later mprotected with PROT_EXEC - the mmaps are all with MAP_ANONYMOUS but I'm not sure we can help it (according to the mprotect manual mprotect() is unspecified if it is applied to a of memory that was not obtained via mmap())

All memory in the seastar allocator is obtained via `mmap()`. So we should be able to mprotect memory coming from it. But we have to be very careful to make sure the result of `mprotect` doesn't spill over to memory used by other objects and to remove it after not needed.